Magento2/Sniffs/PHP/AutogeneratedClassNotInConstructorSniff.php

```diff
@@ -0,0 +1,200 @@
+                if ($variableName === '$this') {
+                    $variableName = $this->getNext($phpcsFile, $variable, $statementEnd, T_STRING)['content'];
+                }
+                foreach ($this->constructorParameters as $parameter) {
+                    $parameterName = $parameter['name'];
+                    if ($this->variableName($parameterName) === $this->variableName($variableName)) {
+                        $variableInParameters = true;
+                    }
+                }
+            }
+
+            if (!$variableInParameters) {
+                $next = $stackPtr;
+                while ($next = $phpcsFile->findNext(T_DOUBLE_COLON, $next + 1, $statementEnd)) {
+                    if ($this->getNext($phpcsFile, $next, $statementEnd, T_STRING)['content'] === 'class') {
+                        $className = $this->getPrevious($phpcsFile, $next, T_STRING)['content'];
+                    }
+                }
+
+                $className = $this->getClassNamespace($className);
+
+                $phpcsFile->addWarning(
+                    sprintf("Class %s needs to be requested in constructor, " .
+                        "otherwise compiler will not be able to find and generate these classes", $className),
+                    $stackPtr,
+                    self::ERROR_CODE
+                );
+            }
+        }
+    }
+
+    /**
+     * Check if it is a ObjectManager::getInstance
+     *
+     * @param File $phpcsFile
+     * @param int $stackPtr
+     * @return bool
+     */
+    private function isObjectManagerGetInstance(File $phpcsFile, int $stackPtr): bool
+    {
+        $prev = $phpcsFile->findPrevious(T_STRING, $stackPtr - 1);
+        $next = $phpcsFile->findNext(T_STRING, $stackPtr + 1);
+        if ($prev &&
+            $next &&
+            $phpcsFile->getTokens()[$prev]['content'] === 'ObjectManager' &&
+            $phpcsFile->getTokens()[$next]['content'] === 'getInstance') {
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Get the complete class namespace from the use's
+     *
+     * @param string $className
+     * @return string
+     */
+    private function getClassNamespace(string $className): string
+    {
+        foreach ($this->uses as $use) {
+            if (end($use) === $className) {
+                $className = implode('/', $use);
+            }
+        }
+        return $className;
+    }
+
+    /**
+     * Register php uses
+     *
+     * @param File $phpcsFile
+     * @param int $stackPtr
+     */
+    private function registerUses(File $phpcsFile, int $stackPtr): void
+    {
+        $useEnd = $phpcsFile->findEndOfStatement($stackPtr);
+        $use = [];
+        $usePosition = $stackPtr;
+        while ($usePosition = $phpcsFile->findNext(T_STRING, $usePosition + 1, $useEnd)) {
+            $use[] = $phpcsFile->getTokens()[$usePosition]['content'];
+        }
+        $this->uses [] = $use;
```

Comment:
It will be easier to process use in `getClassNamespace` if uses will be stored this way
```suggestion
        $this->uses[end($use)] = implode('\', $use);
```

Reply:
This option is easier to compare on line 122.